# alorenze 202008

# description #####
# this script will load count files
# generated by runKallisto pipeline
# and run differential expression
# analysis using DESeq2

# loading packages  ################################################ 
source("scripts/loadingLibs.R")

# setting thresholds ######
padjthreshold = 0.05
log2fcthreshold = 0.75
borderlinezero = 0.25
tpmthreshold = 3

# volcano plot function ################################################ 
volcaPlot = function(allTable, sigTable){
  # Volcano plot with plotly
  ax = list(
    title = "Log_2(FC)",
    zeroline = TRUE,
    showline = TRUE,
    showticklabels = TRUE,
    showgrid = TRUE
  )
  
  ay = list(
    title = "-Log_10(Adjusted p-value)",
    zeroline = TRUE,
    showline = TRUE,
    showticklabels = TRUE,
    showgrid = TRUE
  )
  
  # if there are no DE genes we don't have a way to plot them
  # so this chunk will handle it
  if(dim(sigTable)[1] == 0){
    plot_ly(allTable,
            x = allTable$log2FoldChange,
            y = -log10(allTable$padj),
            text=allTable$target_id,
            type="scatter", mode="markers") %>%
      add_trace(x = log2fcthreshold, mode="lines", text=paste("Log_2(FC) =",log2fcthreshold),
                line = list(color = "black", width = 1)) %>%
      add_trace(x = -log2fcthreshold, mode="lines", text=paste("Log_2(FC) =",-log2fcthreshold),
                line = list(color = "black", width = 1)) %>%
      add_trace(y = -log10(padjthreshold), mode="lines", text=paste("Adjusted p-value =", padjthreshold),
                line = list(color = "black", width = 1)) %>%
      add_trace(y = -log10(padjthreshold), mode="lines", text=paste("Adjusted p-value =", padjthreshold),
                line = list(color = "black", width = 1)) %>%
      layout(showlegend=FALSE, xaxis = ax, yaxis = ay)
  }else{
    plot_ly(allTable,
            x = allTable$log2FoldChange,
            y = -log10(allTable$padj),
            text=allTable$target_id,
            type="scatter", mode="markers") %>%
      add_markers(x = sigTable$log2FoldChange,
                  y = -log10(sigTable$padj),
                  text=sigTable$target_id,
                  mode="markers", type="scatter",
                  marker=list(color="red")) %>%
      add_trace(x = log2fcthreshold, mode="lines", text=paste("Log_2(FC) =",log2fcthreshold),
                line = list(color = "black", width = 1)) %>%
      add_trace(x = -log2fcthreshold, mode="lines", text=paste("Log_2(FC) =",-log2fcthreshold),
                line = list(color = "black", width = 1)) %>%
      add_trace(y = -log10(padjthreshold), mode="lines", text=paste("Adjusted p-value =", padjthreshold),
                line = list(color = "black", width = 1)) %>%
      add_trace(y = -log10(padjthreshold), mode="lines", text=paste("Adjusted p-value =", padjthreshold),
                line = list(color = "black", width = 1)) %>%
      layout(showlegend=FALSE, xaxis = ax, yaxis = ay)
  }
}

# loading raw counts ################################################ 
# to manually compute TPMs, follow the instructions on the following
# page: https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/
# reading totalrna counts
totrna=read_delim("data/tableEstCountsTotalRNA23.tsv",delim="\t")

# reading riboseq counts
riborna=read_delim("data/tableEstCountsRiboSeqTrim15.tsv",delim="\t")

# processing raw counts with DESeq2  ##########################################
## building se object for both
# totrna
samples = totrna %>% 
  dplyr::select(starts_with("total")) %>%
  colnames()
tp = samples %>% 
  sub("^.*-RNA-[1-3]-([1-4])_S.*$", "\\1", .)
bioRep = samples %>% 
  sub("^.*-RNA-([1-3])-[1-4]_S.*$", "\\1", .)

colData = data.frame(row.names = samples,
                     timepoint = tp,
                     replicate = bioRep)

assay = dplyr::select(totrna, starts_with("total")) %>%
  as.matrix() %>% 
  round(digits = 0)

totrnaSE = SummarizedExperiment(assay = list(counts=assay),
                                rowData = totrna[,c(1:3)],
                                colData = colData)
rownames(totrnaSE) = rowData(totrnaSE)$target_id

# riborna
samples = riborna %>% 
  dplyr::select(starts_with("ribosomal")) %>%
  colnames()
tp = samples %>% 
  sub("^.*_RNA_[1-3]-([1-4])_S.*$", "\\1", .)
bioRep = samples %>% 
  sub("^.*_RNA_([1-3])-[1-4]_S.*$", "\\1", .)

colData = data.frame(row.names = samples,
                     timepoint = tp,
                     replicate = bioRep)

assay = dplyr::select(riborna, starts_with("ribosomal")) %>%
  as.matrix() %>% 
  round(digits = 0)

ribornaSE = SummarizedExperiment(assay = list(counts=assay),
                                rowData = riborna[,c(1:3)],
                                colData = colData)
rownames(ribornaSE) = rowData(ribornaSE)$target_id

# creating deseq2 objects
totrnadds = totrnaSE
totrnadds = DESeqDataSet(totrnadds, design = ~ timepoint)

ribodds = ribornaSE
ribodds = DESeqDataSet(ribodds, design = ~ timepoint)

# removing genes with zero counts and performing DESeq2 analysis
totrnadds = totrnadds[rowSums(counts(totrnadds)) > 1, ]
totrnadds = DESeq(totrnadds)

ribodds = ribodds[rowSums(counts(ribodds)) > 1, ]
ribodds = DESeq(ribodds)

# result tables for contrasts
results = list()
for(type in c("totrna", "ribo")){
  for(i in 2:4 %>% as.character()){
    results[[paste0(type,i)]] = results(get(paste0(type,"dds")),
                                        contrast= c("timepoint", i, "1"),
                                        alpha = padjthreshold) %>% 
      as_tibble(rownames = "target_id")
  }
}

# result sig tables for contrasts
# only to plot volcanos
resultsSig = list()
for(type in c("totrna", "ribo")){
  for(i in 2:4 %>% as.character()){
    resultsSig[[paste0(type,i)]] = results[[paste0(type,i)]] %>%
      dplyr::filter(abs(log2FoldChange) >= log2fcthreshold & padj < padjthreshold) %>% 
      dplyr::filter(padj < padjthreshold)
  }
}

# volcano plot
# for(type in c("totrna", "ribo")){
#   for(i in 2:4 %>% as.character()){
#     p = volcaPlot(results[[paste0(type,i)]], resultsSig[[paste0(type,i)]])
#     htmlwidgets::saveWidget(widget = p, file = paste0("plots/", type, i, "_deseq2.html"))
#   }
# }

# final result tables
# containing sig status and borderline status
resultsFin = list()
for(type in c("totrna", "ribo")){
  for(i in 2:4 %>% as.character()){
    varName = paste0("sig",type,"_TP",i)
    varName2 = paste0("borderlineZero",type,"_TP",i)
    resultsFin[[paste0(type,i)]] = results[[paste0(type,i)]] %>% 
      mutate(!!varName := case_when(abs(log2FoldChange) >= log2fcthreshold & padj < padjthreshold ~ "yes",
                                    TRUE ~ "no")) %>% 
      mutate(!!varName2 := case_when(abs(log2FoldChange) < borderlinezero ~ "yes", 
                                     TRUE ~ "no"))
  }
}

# unifying final results
unifiedFin = list()
for(i in 2:4 %>% as.character()){
  unifiedFin[[paste0("TP",i)]] = inner_join(resultsFin[[paste0("totrna",i)]],
                                            resultsFin[[paste0("ribo",i)]],
                                            by = "target_id") %>% 
    rename("id" = target_id) %>% 
    mutate(locus_tag = sub("\\|.*$", "", id)) %>% 
    dplyr::select(locus_tag,
           everything(),
           -id) %>% 
    rename_with(.fn = ~ str_replace(string = .x,
                                 pattern = "_TP.*$",
                                 replacement = ""),
                .cols = matches("TP")) %>% 
    rename("mean_lfc_rna_total" = log2FoldChange.x,
           "se_lfc_rna_total" = lfcSE.x,
           "mean_lfc_rna_ribo" = log2FoldChange.y,
           "se_lfc_rna_ribo" = lfcSE.y)
}

